cachi2/core/package_managers/yarn/project.py

```diff
@@ -120,9 +148,12 @@ def yarn_cache(self) -> RootedPath:
         The cache location is affected by the cacheFolder configuration in yarnrc. See:
         https://v3.yarnpkg.com/configuration/yarnrc#cacheFolder.
         """
-        return NotImplemented
+        return self.source_dir.join_within_root(self.yarn_rc.cache_path)
 
     @classmethod
     def from_source_dir(cls, source_dir: RootedPath) -> "Project":
         """Create a Project from a sources directory path."""
-        return NotImplemented
+        yarn_rc = YarnRc.from_file(source_dir.join_within_root(".yarnrc.yml"))
+        package_json = PackageJson.from_file(source_dir.join_within_root("package.json"))
```

Comment:
The [docs](https://v3.yarnpkg.com/configuration/yarnrc) state that:
```
While Yarn will automatically find them in the parent directories, they should usually be kept at the root of your project (often your repository)
```

So that is one thing we need to account for when covering the path/symlink validation story.

Reply:
`source_dir.join_within_root("package.json")` does the validation right?

Reply:
I was thinking that a parent dir package.json could be accessed by yarn itself when we invoked any of the yarn commands. But with the validation to enforce that a valid package.json file exists in every project, this will not happen.